homeassistant/components/switch/rainmachine.py

```diff
@@ -0,0 +1,308 @@
+"""Implements a RainMachine sprinkler controller for Home Assistant."""
+
+import asyncio
+from datetime import timedelta
+from logging import getLogger
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.switch import SwitchDevice, PLATFORM_SCHEMA
+from homeassistant.const import (ATTR_ATTRIBUTION, ATTR_DEVICE_CLASS,
+                                 CONF_EMAIL, CONF_IP_ADDRESS, CONF_PASSWORD)
+from homeassistant.util import Throttle
+
+_LOGGER = getLogger(__name__)
+REQUIREMENTS = ['regenmaschine==0.3.2']
+
+ATTR_CYCLES = 'cycles'
+ATTR_TOTAL_DURATION = 'total_duration'
+
+CONF_HIDE_DISABLED_ENTITIES = 'hide_disabled_entities'
+CONF_ZONE_RUN_TIME = 'zone_run_time'
+
+DEFAULT_ZONE_RUN_SECONDS = 60 * 10
+
+MIN_SCAN_TIME_LOCAL = timedelta(seconds=1)
+MIN_SCAN_TIME_REMOTE = timedelta(seconds=5)
+MIN_SCAN_TIME_FORCED = timedelta(milliseconds=100)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_IP_ADDRESS):
```

Comment:
Require at least on of the auth keys and make them exclusive, something like this:
```py
PLATFORM_SCHEMA = vol.Schema(vol.All(
    cv.has_at_least_one_key(CONF_IP_ADDRESS, CONF_EMAIL),
    {
        vol.Required(CONF_PLATFORM): cv.string,
        vol.Optional(CONF_SCAN_INTERVAL): cv.time_period
        vol.Exclusive(CONF_IP_ADDRESS, 'auth'): cv.string,
        vol.Exclusive(CONF_EMAIL, 'auth'): vol.Email,
        vol.Required(CONF_PASSWORD): cv.string,
        vol.Optional(CONF_ZONE_RUN_TIME, default=DEFAULT_ZONE_RUN_SECONDS): cv.positive_int,
        vol.Optional(CONF_HIDE_DISABLED_ENTITIES, default=True): cv.boolean
    }
), extra=vol.ALLOW_EXTRA)
```
Note that you can't extend the platform schema when you want to use a validator around the dict.